smart-contracts/contracts/mixins/MixinKeys.sol

```diff
@@ -340,8 +340,16 @@ contract MixinKeys is
   ) internal
   {
     if (ownerOf(_tokenId) != _keyOwner) {
-      // TODO: this may include duplicate entries
-      owners.push(_keyOwner);
+      // make sure we don't add to owners list if a key was previously owned
+      bool isAlreadyRecorded = false;
+      for (uint256 i = 0; i < owners.length; i++) {
```

Comment:
this is the most straightforward solution to the issue but obviously it could be quite gas intensive